### PR TITLE
dim-web

### PR DESCRIPTION
**Pull request by mental32**
Porting the webserver from warp to axum.

All of the routes are still using warp, however they are now turned into `tower::Service`s and are all using axum's router abstraction + route_service.

Due to how websockets work we could not do the same with the /ws endpoint, so this endpoint has been the only one completely migrated to axum.

### CODE REVIEW

File: dim/Cargo.toml [hunk 1]
```diff
@@ -60,15 +63,18 @@ uuid = { version = "1.2.2", features = ["v4"] }
 walkdir = "2.3.1"
 warp = { version = "0.3.3", features = ["tls", "tokio-rustls"] }
 xmlwriter = "0.1.0"
-xtra = { version = "0.5.1", features = ["with-tokio-1"] }
+xtra = { version = "0.5.1", features = ["tokio", "with-tokio-1"] }
 
 # FIXME: Remove when we get rid of xtra_proc
-new_xtra = { package = "xtra", git = "https://github.com/Restioson/xtra", features = ["with-tokio-1"] }
+# new_xtra = { package = "xtra", git = "https://github.com/Restioson/xtra", features = ["tokio"] }
```
Comment (vgarleanu) [inline dim/Cargo.toml:69]: Should probably just remove it altogether

File: dim/src/core.rs [hunk 2]
```diff
@@ -95,115 +100,308 @@ pub async fn warp_core(
...
+            warp!(routes::general::filters::get_directory_structure),
+        )
+        //     routes::general::filters::get_directory_structure(conn.clone()),
+        .route_service("/images/*path", warp!(statik::filters::get_image))
+        //     routes::statik::filters::get_image(conn.clone()),
+        //     /* media routes */
+        .route_service(
+            "/api/v1/media/*path",
+            warp::service({
+                //     routes::media::filters::get_media_by_id(conn.clone()),
```
Comment (vgarleanu) [inline dim/src/core.rs:210]: Do we need all of these comments merged too?

File: dim/src/core.rs [hunk 3]
```diff
@@ -95,115 +100,308 @@ pub async fn warp_core(
...
+                routes::statik::filters::dist_static().or(routes::statik::filters::react_routes()),
+            ),
+        )
+        .route("/ws", dim_web::axum::routing::get(ws_handler))
+        .with_state(AppState {
+            conn: conn.clone(),
+            socket_tx,
+        })
+        .layer(tower_http::trace::TraceLayer::new_for_http())
+        .layer({
```
Comment (vgarleanu) [inline dim/src/core.rs:389]: Don't think we need an explicit block.

File: dim/src/routes/dashboard.rs [hunk 4]
```diff
@@ -25,24 +25,18 @@ pub mod filters {
...
-
     pub fn dashboard(
         conn: DbConnection,
-        rt: tokio::runtime::Handle,
     ) -> impl Filter<Extract = impl warp::Reply, Error = warp::Rejection> + Clone {
         warp::path!("api" / "v1" / "dashboard")
             .and(warp::get())
             .and(with_auth(conn.clone()))
             .and(with_state::<DbConnection>(conn))
-            .and(with_state::<TokioHandle>(rt))
```
Comment (vgarleanu) [inline dim/src/routes/dashboard.rs:38]: Can you manually run cargo fmt pls. Doesn't seem like vscode is picking up the rustfmt file

File: dim/src/routes/mod.rs [hunk 5]
```diff
@@ -41,14 +42,19 @@ pub mod global_filters {
...
     use warp::Reply;
 
-    pub fn with_db(
-        conn: DbConnection,
-    ) -> impl Filter<Extract = (DbConnection,), Error = Infallible> + Clone {
+    pub fn with_db<C>(
+        conn: C,
+    ) -> impl Filter<Extract = (DbConnection,), Error = Infallible> + Clone + 'static
+    where
+        C: Borrow<DbConnection>,
```
Comment (vgarleanu) [inline dim/src/routes/mod.rs:55]: Why the borrow?

**Review by vgarleanu — CHANGES_REQUESTED** [review]

File: dim/src/routes/websocket.rs [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+            None = stream.next() => {
+                let _ = socket_tx.send(CtrlEvent::Forget { addr });
+                break;
+            }
+        }
+    }
+}
+
+fn from_warp_message(inner: warp::ws::Message) -> Result<WsMessage, ()> {
+    if inner.is_binary() {
```
[the remark this answers is not in the document]
Comment (mental32) [reply dim/src/routes/websocket.rs]: nope

File: dim/src/core.rs (hunk 2 above)
Comment (mental32) [reply dim/src/core.rs:210]: I'll remove them

File: dim/src/core.rs (hunk 3 above)
Comment (mental32) [reply dim/src/core.rs:389]: i think it's fine for now, these handers are getting tossed in the next couple of PRs anyway

File: dim/src/routes/mod.rs (hunk 5 above)
Comment (mental32) [reply dim/src/routes/mod.rs:55]: I wanted to see if i didn't have to write `conn.clone()` everywhere

File: dim/src/routes/websocket.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (vgarleanu) [reply dim/src/routes/websocket.rs]: Seems like you're doing it on like 219 tho

[the remark this answers is not in the document]
Comment (mental32) [reply dim/src/routes/websocket.rs]: it's not an enum its a struct that wraps the enum, those are constructor functions for the variants.

File: dim-web/src/routes/websocket.rs [hunk 7]
```diff
@@ -0,0 +1,259 @@
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::net::SocketAddr;
+use std::pin::Pin;
+
+use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
```
Comment (vgarleanu) [inline dim-web/src/routes/websocket.rs:6]: For style I prefer the imports to be broken down.
```suggestion
use tokio::sync::mpsc::unbounded_channel;
use tokio::sync::mpsc::UnboundedReceiver;
use tokio::sync::mpsc::UnboundedSender;
```

**Review by vgarleanu — APPROVED** [review]

**Pull request merged**